### PR TITLE
[youtube:comments] Add more options for limiting comments extracted and misc cleanup

### PR DESCRIPTION
**Pull request by coletdjnz**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [X] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [X] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [X] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [X] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [X] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

This adds more options for limiting the amount of comments retrieved for youtube, as requested in #1390

`youtube:max_comments=max-comments,max-parents,max-replies,max-replies-per-thread`
This also cleans up old deprecated comment extraction code that is not used anymore.
Also removed the code that stops if no comments are received - appears that YouTube have fixed this issue their end. It was also causing a premature exit in some cases. 

TODO:
- [X] Remove deprecated extraction code
- [X] Add more extractor args for limiting comments
- [X] Remove stop if no comments received  
- [X] Fix up or check visitor data extraction is working correctly
- [X] Fix up the messy code we just added
- [X] Testing

### CODE REVIEW

File: yt_dlp/extractor/youtube.py [hunk 1]
```diff
@@ -2091,8 +2092,17 @@ def _extract_comment(self, comment_renderer, parent=None):
...
         }
 
-    def _comment_entries(self, root_continuation_data, ytcfg, video_id, parent=None, comment_counts=None):
+    @dataclasses.dataclass
+    class _CommentTracker:
+        running_total: int = 0
+        est_total: int = 0
+        current_page_thread: int = 0
+        total_parent_comments: int = 0
+        total_reply_comments: int = 0
```
Comment (coletdjnz) [inline yt_dlp/extractor/youtube.py:2101]: Anyone know of a better way of going about this? Before we were using a list and it wasn't easily readable. The recursive design of the comment extractor makes keeping track of this data a bit messy.

File: yt_dlp/extractor/youtube.py [hunk 2]
```diff
@@ -2220,7 +2228,11 @@ def extract_thread(contents):
                             break
                         continue
                     count = 0
+                    max_parent_comments = int_or_none(get_single_config_arg('max_parent_comments')) or float('inf')
+                    max_reply_comments = int_or_none(get_single_config_arg('max_reply_comments')) or float('inf')
                     for count, entry in enumerate(extract_thread(continuation_items)):
+                        if comment_tracker.total_parent_comments >= max_parent_comments or comment_tracker.total_reply_comments >= max_reply_comments:
+                            return
```
Comment (coletdjnz) [inline yt_dlp/extractor/youtube.py:2235]: this could possibly go in the `_get_comments` function

File: yt_dlp/extractor/youtube.py (hunk 1 above)
Comment (coletdjnz) [reply yt_dlp/extractor/youtube.py:2101]: darn, can't use this anyways as dataclasses are not supported in py3.6

Comment (pukkandan) [reply yt_dlp/extractor/youtube.py:2101]: only way I can think of is to put the entire comment extraction code in a class/closure

Comment (coletdjnz) [reply yt_dlp/extractor/youtube.py]: Yeah :/
maybe that can be a future improvement

File: yt_dlp/extractor/youtube.py [hunk 3]
```diff
@@ -2134,76 +2133,84 @@ def extract_header(contents):
...
                         yield reply_comment
 
+        # Keeps track of counts across recursive calls
+        # 0: running_total
+        # 1: est_total
+        # 2: current_page_thread
+        # 3: total_parent_comments
+        # 4: total_reply_comments
+        # TODO: implement a cleaner / more maintainable method?
+        comment_tracker = comment_tracker or [0, 0, 0, 0, 0]
```
Comment (pukkandan) [inline yt_dlp/extractor/youtube.py:2188]: using a namedtuple/dict would make this self-documenting

File: yt_dlp/extractor/youtube.py [hunk 4]
```diff
@@ -2212,83 +2219,33 @@ def extract_thread(contents):
...
-                    count = 0
-                    for count, entry in enumerate(extract_thread(continuation_items)):
-                        yield entry
-                    continuation = self._extract_continuation({'contents': continuation_items})
+            for continuation_section in continuation_contents:
+                continuation_items = try_get(
+                    continuation_section,
+                    (lambda x: x['reloadContinuationItemsCommand']['continuationItems'],
+                     lambda x: x['appendContinuationItemsAction']['continuationItems']),
+                    list) or []
```
Comment (pukkandan) [inline yt_dlp/extractor/youtube.py:2233]: pretty sure this is same as:
```suggestion
                continuation_items = get_first(
                    continuation_section,
                    ('reloadContinuationItemsCommand', 'appendContinuationItemsAction'), 'continuationItems',
                    expected_type=list) or []
```

Comment (coletdjnz) [reply yt_dlp/extractor/youtube.py:2233]: hmm not quite

```suggestion
                continuation_items = get_first(
                    continuation_section,
                    [[('reloadContinuationItemsCommand', 'appendContinuationItemsAction'), 'continuationItems']],
                    expected_type=list) or []
```
but this seems to work?

Comment (pukkandan) [reply yt_dlp/extractor/youtube.py:2233]: hm.. I though `get_first` takes `*args`, but seems not
```py
def get_first(obj, keys, **kwargs):
    return traverse_obj(obj, (..., *variadic(keys)), **kwargs, get_all=False)
```
So it should be `(('reloadContinuationItemsCommand', 'appendContinuationItemsAction'), 'continuationItems')`. Are you sure there needs to be 2 levels of nesting?

Comment (coletdjnz) [reply yt_dlp/extractor/youtube.py:2233]: `(('reloadContinuationItemsCommand', 'appendContinuationItemsAction'), 'continuationItems')` doesn't work either. 

But with two levels of nesting it seems to work

Comment (pukkandan) [reply yt_dlp/extractor/youtube.py:2233]: sorry, get_first isn't suitable here. I was thinking directly in terms of `traverse_obj`
```
>>> from yt_dlp.utils import traverse_obj
>>> d = {'x': {'y': 'z'}}
>>> traverse_obj(d, (('x', 'a'), 'y'), get_all=False)
'z'
```
So:
```suggestion
                continuation_items = traverse_obj(
                    continuation_section,
                    (('reloadContinuationItemsCommand', appendContinuationItemsAction'), 'continuationItems'),
                    get_all=False, expected_type=list) or []
```

File: README.md [hunk 5]
```diff
@@ -1565,6 +1565,9 @@ The following extractors use this feature:
 * `include_live_dash`: Include live dash formats (These formats don't download properly)
 * `comment_sort`: `top` or `new` (default) - choose comment sorting mode (on YouTube's side)
 * `max_comments`: Maximum amount of comments to download (default all)
+* `max_reply_comments_per_thread`: Maximum amount of reply comments per thread (default all)
+* `max_parent_comments`: Maximum parent comments overall (default all)
+* `max_reply_comments`: Maximum reply comments overall (default all)
 * `max_comment_depth`: Maximum depth for nested comments. YouTube supports depths 1 or 2 (default)
```
[the remark this answers is not in the document]
Comment (coletdjnz) [reply README.md:1571]: > Also, setting depth=1 is same as max_replies=0?
yep. I guess we can deprecate the max_comment_depth option

[the remark this answers is not in the document]
Comment (pukkandan) [reply README.md:1571]: Yeah, dont remove it from code rn. Just remove from docs. Also maybe issue a warning when using it. We can fully remove it later

File: yt_dlp/extractor/youtube.py [hunk 6]
```diff
@@ -2187,11 +2180,15 @@ def extract_thread(contents):
                 total_parent_comments=0,
                 total_reply_comments=0)
 
+        # TODO: Deprecated
         # YouTube comments have a max depth of 2
         max_depth = int_or_none(get_single_config_arg('max_comment_depth')) or 2
         if max_depth == 1 and parent:
```
Comment (pukkandan) [inline yt_dlp/extractor/youtube.py]: Merge with master. Then you can do:
```py
        max_depth = int_or_none(get_single_config_arg('max_comment_depth'))
        if max_depth:
            self._downloader.deprecation_warning('[youtube] ...')
```

**Review by pukkandan — APPROVED** [review]
LGTM. Can merge when ready

**Pull request merged**